resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -525,9 +534,11 @@ private[yarn] class YarnAllocator(
               } catch {
                 case NonFatal(e) =>
                   logError(s"Failed to launch executor $executorId on container $containerId", e)
-                  // Assigned container should be released immediately to avoid unnecessary resource
-                  // occupation.
+                  // Assigned container should be released immediately
+                  // to avoid unnecessary resource occupation.
                   amClient.releaseAssignedContainer(containerId)
+              } finally {
+                numExecutorsStarting.decrementAndGet()
```

Comment:
why did you move this out of updateInternalState?  I would rather see those stay together because updateInternalState is synchronized and if they aren't changed at the same time you could end up with the wrong numbers. numExecutorsStarting could still be say 1 when you have already added that to the # running and then updateResourceRequests gets an incorrect number.